cirq-core/cirq/ops/gateset.py

```diff
@@ -0,0 +1,292 @@
+            b) `cirq.Gate` instance
+            c) `cirq.GateFamily` instance
+
+        `cirq.Gate` subclasses and instances are converted to the default
+        `cirq.GateFamily(gate=g)` instance and thus a default name and
+        description is populated.
+
+        Args:
+            gates: A list of `cirq.Gate` subclasses / `cirq.Gate` instances /
+            `cirq.GateFamily` instances to initialize the Gateset.
+            name: (Optional) Name for the Gateset. Useful for description.
+        """
+        self._name = name
+        self._gates = frozenset(
+            [g if isinstance(g, GateFamily) else GateFamily(gate=g) for g in gates]
+        )
+
+    @property
+    def name(self) -> Optional[str]:
+        return self._name
+
+    @property
+    def gates(self) -> FrozenSet[GateFamily]:
+        return self._gates
+
+    def __contains__(self, item: Union[raw_types.Gate, raw_types.Operation]) -> bool:
+        """Check for containment of `item` in any of the member `GateFamily`.
+
+        This method just forwards containment checks to the underlying
+        `GateFamily` objects. It returns `True` if any of the member gate
+        families can accept `item`.
+
+        To validate whether a `cirq.CircuitOperation`/`cirq.OP_TREE`/
+        `cirq.Circuit` is made up entirely of gates from this Gateset, use the
+        `validate` method given below.
+
+        Args:
+            item: The `cirq.Gate` or `cirq.Operation` instance to check
+                containment for.
+        """
+        for gate_family in self.gates:
+            if item in gate_family:
+                return True
+        return False
+
+    def validate(
+        self,
+        circuit_or_optree: Union['cirq.AbstractCircuit', op_tree.OP_TREE],
+        *,
+        unroll_circuit_op: bool = True,
+        accept_global_phase: bool = True,
+    ) -> bool:
+        """Validates gates forming `circuit_or_optree` should be contained in Gateset.
+
+        Args:
+            circuit_or_optree: The `cirq.Circuit` or `cirq.OP_TREE` to validate.
+            unroll_circuit_op: If True, `cirq.CircuitOperation` is recursively
+                validated by validating the underlying `cirq.Circuit`.
+            accept_global_phase: If True, `cirq.GlobalPhaseOperation` is accepted.
+        """
+        # To avoid circular import.
+        from cirq.circuits import circuit
+
+        optree = circuit_or_optree
+        if isinstance(circuit_or_optree, circuit.AbstractCircuit):
+            optree = circuit_or_optree.all_operations()
+        return all(
+            self._validate_operation(
+                op, unroll_circuit_op=unroll_circuit_op, accept_global_phase=accept_global_phase
+            )
+            for op in op_tree.flatten_to_ops(optree)
+        )
+
+    def _validate_operation(
+        self,
+        op: raw_types.Operation,
+        *,
+        unroll_circuit_op: bool = True,
+        accept_global_phase: bool = True,
```

Comment:
The idea is to keep `__contains__` simple enough by delegating the containment checks to the underlying `GateFamily` and having additional helpers like `validate` which try to be smart while handling different types of operations (eg: `GlobalPhaseOperation`, `CircuitOperation` etc.). This is also why `contains` doesn't provide ways to deal with other types of operations like `CircuitOperations`.

I think it's beneficial to keep the smart handling of different operation types decoupled from the basic functionality of checking for containment in the underlying gate families.

Reply:
That makes sense to me.

Thinking more about this I'm wondering if we can move both: `unroll_circuit_op` and `accept_global_phase` global phase into the constructor. If a user is a given a gateset to work with, they may not immediately know which values these flags should take on in their particular setting and it would be better to fully specify the kinds of support this gateset has during construction.

Reply:
Hmm, that's a good point. How about having the default values of these flags as part of the constructor but also allowing an override of the default values in `validate` ? I think we should allow the flexibility of validating circuits with these options configurable.

Reply:
I think it would make more sense to just have it all be in the constructor. Giving the user more than one way to specify these options that looks very similar but behave slightly differently just seems like it's giving more ways for things to go wrong in a confusing way.

Reply:
done for now. We can add params to `validate` if we have a strong use case later.